modules/metastorage-server/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/WatchCursor.java

```diff
@@ -0,0 +1,234 @@
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Nullable
+    @Override public WatchEvent next() {
+        return it.next();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override public void close() throws Exception {
+        IgniteUtils.closeAll(options, nativeIterator);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @NotNull
+    @Override public Iterator<WatchEvent> iterator() {
+        return it;
+    }
+
+    /**
+     * Creates an iterator for this cursor.
+     *
+     * @return Iterator.
+     */
+    @NotNull
+    private Iterator<WatchEvent> createIterator() {
+        return new Iterator<>() {
+            /** {@inheritDoc} */
+            @Override public boolean hasNext() {
+                storage.lock().readLock().lock();
+
+                try {
+                    if (nextRetRev != -1)
+                        // Next revision is already calculated and is not -1, meaning that there is set of keys
+                        // matching the revision and the predicate.
+                        return true;
+
+                    while (true) {
+                        long curRev = lastRetRev + 1;
+
+                        byte[] revisionPrefix = ByteUtils.longToBytes(curRev);
+
+                        boolean empty = true;
+
+                        if (!nativeIterator.isValid())
+                            try {
+                                nativeIterator.refresh();
+                            }
+                            catch (RocksDBException e) {
+                                throw new IgniteInternalException(e);
+                            }
+
+                        // Check all keys by the revision to see if any one of them match the predicate.
+                        for (nativeIterator.seek(revisionPrefix); nativeIterator.isValid(); nativeIterator.next()) {
+                            empty = false;
+
+                            byte[] key = RocksDBKeyValueStorage.rocksKeyToBytes(nativeIterator.key());
+
+                            if (p.test(key)) {
+                                // Current revision matches.
+                                nextRetRev = curRev;
+
+                                return true;
+                            }
+                        }
+
+                        RocksDBKeyValueStorage.checkIterator(nativeIterator);
+
+                        if (empty)
+                            return false;
+
+                        // Go to the next revision.
+                        lastRetRev++;
+                    }
+                }
+                finally {
+                    storage.lock().readLock().unlock();
+                }
+            }
+
+            /** {@inheritDoc} */
+            @Override public WatchEvent next() {
+                storage.lock().readLock().lock();
+
+                try {
+                    while (true) {
+                        if (!hasNext())
+                            return null;
```

Comment:
Ok, but I doubt that it will make any difference since this method is declared in the `Iterator` class. There will be no place to check the invocation of this specific implementation.

Reply:
yeah, I understand, it's just so that IDEA will not complain